packages/evm/contracts/adapters/Proofcast/ProofcastAdapter.sol

```diff
@@ -0,0 +1,122 @@
+            emit TeeSignerPendingChange(teeAddressNew, attestation, teeAddressChangeGraceThreshold);
+        }
+    }
+
+    function storeHashes(bytes calldata statement, bytes memory signature) public {
+        if (teeAddressNew != address(0) && block.timestamp > teeAddressChangeGraceThreshold) {
+            teeAddress = teeAddressNew;
+            teeAddressNew = address(0);
+            emit TeeSignerChanged(teeAddress);
+        }
+        require(teeAddress != address(0), "Tee signer not set!");
+        require(ECDSA.recover(sha256(statement), signature) == teeAddress, "Invalid signature");
+
+        // Supports only EVM events
+        bytes1 protocolId = statement[1];
+        if (protocolId != 0x00) revert UnsupportedProtocolId(protocolId);
+
+        (uint256 domain, uint256[] memory ids, bytes32[] memory hashes) = _checkEventAndDecodeData(statement);
+        _storeHashes(domain, ids, hashes);
+    }
+
+    function _checkEventAndDecodeData(
+        bytes calldata statement
+    ) internal view returns (uint256 domain, uint256[] memory ids, bytes32[] memory hashes) {
+        //  Statement format:
+        //    | version   | protocol   |  protocol_chain_id |   event id    | event_bytes |
+        //    | (1 byte)  | (1 byte)   |    (32 bytes)      |  (32 bytes)   |  varlen     |
+
+        uint16 offset = 2; // skip version, protocolId
+        domain = uint256(bytes32(statement[offset:(offset += 32)]));
+
+        if (yahos[domain] == address(0)) revert UnsupportedChainId(domain);
+
+        offset += 32; // skip the event id (32 bytes)
+        bytes memory eventBytes = statement[offset:];
+        RLPReader.RLPItem memory eventRLP = RLPReader.toRlpItem(eventBytes);
+        if (!RLPReader.isList(eventRLP)) revert InvalidEventRLP();
+
+        RLPReader.RLPItem[] memory eventContent = RLPReader.toList(eventRLP);
+
+        // Event must contain address, logs and data
+        if (eventContent.length != 3) revert InvalidEventContentLength(eventContent.length);
+
+        // MessageDispatched event parsing
+        address yahoAddress = RLPReader.toAddress(eventContent[0]);
+        require(yahoAddress == yahos[domain], "Invalid Yaho address");
+
+        RLPReader.RLPItem[] memory logs = RLPReader.toList(eventContent[1]);
+
+        bytes32 topic = bytes32(RLPReader.toBytes(logs[0]));
+        if (topic != MESSAGE_DISPATCHED_EVENT_TOPIC) revert UnexpectedEventTopic(topic);
+
+        bytes memory messageBytes = RLPReader.toBytes(eventContent[2]);
+        Message memory message = abi.decode(messageBytes, (Message));
+        uint256 expectedMessageId = calculateMessageId(domain, yahoAddress, calculateMessageHash(message));
+
+        uint256 messageId = uint256(bytes32(RLPReader.toBytes(logs[1])));
+        if (messageId != expectedMessageId) revert InvalidMessageId(messageId, expectedMessageId);
+
+        (ids, hashes) = abi.decode(message.data, (uint256[], bytes32[]));
```

Comment:
`message.data` doesn't contain the `messageIds` and the `hashes`. `message.data` is just the abitrary data that is sent cross-chain. 

```sol
messageHash = calculateMessageHash(message)
messageId = calculateMessageId(chainId, yahoAddress, messageHash)
ids[0] = messageId
hashes[0] = messageHash
```